### PR TITLE
Fix(ServiceClient): Fix issue with IOThreadTimer nullref.

### PR DESCRIPTION
**Pull request by azabbasi**
If the IOThreadTimer is canceled twice without being set in between the cancels, it will throw a nullref because there is no null check involved.

### CODE REVIEW

File: iothub/service/src/IotHubConnection.cs [hunk 1]
```diff
@@ -47,14 +47,15 @@ internal sealed class IotHubConnection : IDisposable
 
         public IotHubConnection(IotHubConnectionString connectionString, AccessRights accessRights, bool useWebSocketOnly, ServiceClientTransportSettings transportSettings)
         {
-            ConnectionString = connectionString;
-            _accessRights = accessRights;
-            _faultTolerantSession = new FaultTolerantAmqpObject<AmqpSession>(CreateSessionAsync, CloseConnection);
 #if !NET451
-            _refreshTokenTimer = new IOThreadTimerSlim(s => ((IotHubConnection)s).OnRefreshToken(), this, false);
+            _refreshTokenTimer = new IOThreadTimerSlim(s => ((IotHubConnection)s).OnRefreshTokenAsync(), this);
```
Comment (drwill-ms) [inline iothub/service/src/IotHubConnection.cs:51]: was the change of the `false` parameter from !NET451 to else intentional? Can you say more about why it was wrong?

File: iothub/service/src/IotHubConnection.cs [hunk 2]
```diff
@@ -428,17 +422,9 @@ private async void OnRefreshToken()
...
-            }
-
-            if (s_disableServerCertificateValidation.Value && sslPolicyErrors == SslPolicyErrors.RemoteCertificateNameMismatch)
-            {
-                return true;
-            }
-
-            return false;
+            return sslPolicyErrors == SslPolicyErrors.None 
+                || (s_disableServerCertificateValidation.Value
```
Comment (drwill-ms) [inline iothub/service/src/IotHubConnection.cs:426]: If `s_disableServerCertificationValidation` is a bool, can we rename it to something that sounds more like a bool?

s_shouldDisableServerCertificationValidation?

File: iothub/service/src/IotHubConnection.cs (hunk 1 above)
Comment (azabbasi) [reply iothub/service/src/IotHubConnection.cs:51]: That parameter was just not used in the IOThreadTimerSlim so I just removed it

File: iothub/service/src/IotHubConnection.cs (hunk 2 above)
Comment (azabbasi) [reply iothub/service/src/IotHubConnection.cs:426]: sure !

File: iothub/service/src/net451/Common/IOThreadTimer.cs [hunk 3]
```diff
@@ -122,6 +122,11 @@ public void SetAt(long newDueTimeInTicks)
             TimerManager.Value.Set(this, newDueTimeInTicks);
         }
 
+        public void Dispose()
```
Comment (drwill-ms) [inline iothub/service/src/net451/Common/IOThreadTimer.cs:125]: good catch

**Review by drwill-ms — APPROVED** [review]

File: iothub/service/src/IotHubConnection.cs (hunk 1 above)
Comment (drwill-ms) [reply iothub/service/src/IotHubConnection.cs:51]: What about line 53 then?

File: iothub/service/src/ModernDotNet/Common/IOThreadTimerSlim.cs [hunk 4]
```diff
@@ -1,43 +1,52 @@
...
+                _timer.Dispose();
+                _timer = null;
+            }
+
             return true;
         }
+
+        public void Dispose()
+        {
+            _timer.Dispose();
```
Comment (drwill-ms) [inline iothub/service/src/ModernDotNet/Common/IOThreadTimerSlim.cs:44]: _timer?.Dispose()

File: iothub/service/src/ModernDotNet/Common/IOThreadTimerSlim.cs [hunk 5]
```diff
@@ -1,43 +1,52 @@
...
-            this.callbackState = callbackState;
+            _callback = callback;
+            _callbackState = callbackState;
             CreateTimer();
         }
 
         public void Set(TimeSpan timeFromNow)
         {
-            if (timer == null)
+            if (_timer == null)
```
Comment (drwill-ms) [inline iothub/service/src/ModernDotNet/Common/IOThreadTimerSlim.cs:24]: Probably should use a lock or a semaphore to control access to _timer

File: iothub/service/src/ModernDotNet/Common/IOThreadTimerSlim.cs [hunk 6]
```diff
@@ -1,43 +1,52 @@
...
             {
                 CreateTimer();
             }
-            timer.Change(timeFromNow, TimeSpan.FromMilliseconds(-1));
+            _timer.Change(timeFromNow, TimeSpan.FromMilliseconds(-1));
```
Comment (drwill-ms) [inline iothub/service/src/ModernDotNet/Common/IOThreadTimerSlim.cs:28]: `TimeSpan.Infinite` for 2nd parameter

File: iothub/service/src/IotHubConnection.cs [hunk 7]
```diff
@@ -76,22 +77,16 @@ public Task CloseAsync()
             return _faultTolerantSession.CloseAsync();
         }
 
-        public void SafeClose(Exception exception)
```
Comment (abhipsaMisra) [inline iothub/service/src/IotHubConnection.cs:79]: was this unused?

Comment (azabbasi) [reply iothub/service/src/IotHubConnection.cs:79]: yes

File: iothub/service/src/IotHubConnection.cs [hunk 8]
```diff
@@ -467,6 +453,7 @@ public static ArraySegment<byte> ConvertToDeliveryTag(string lockToken)
         public void Dispose()
         {
             _faultTolerantSession.Dispose();
+            _refreshTokenTimer.Dispose();
```
Comment (abhipsaMisra) [inline iothub/service/src/IotHubConnection.cs:456]: good catch

File: iothub/service/src/ModernDotNet/Common/IOThreadTimerSlim.cs [hunk 9]
```diff
@@ -1,43 +1,62 @@
...
+        private readonly Action<object> _callback;
+        private readonly object _callbackState;
+        private SemaphoreSlim _sem;
 
-        public IOThreadTimerSlim(Action<object> callback, object callbackState, bool isTypicallyCanceledShortlyAfterBeingSet)
+        public IOThreadTimerSlim(Action<object> callback, object callbackState)
         {
-            this.callback = callback;
-            this.callbackState = callbackState;
+            _sem = new SemaphoreSlim(1);
```
Comment (drwill-ms) [inline iothub/service/src/ModernDotNet/Common/IOThreadTimerSlim.cs:18]: Doesn't this need disposing as well?

Comment (azabbasi) [reply iothub/service/src/ModernDotNet/Common/IOThreadTimerSlim.cs:18]: Yes

File: iothub/service/src/ModernDotNet/Common/IOThreadTimerSlim.cs [hunk 10]
```diff
@@ -1,43 +1,63 @@
...
-        object callbackState;
-
-        private void CreateTimer()
-        {
-            this.timer = new Timer((obj) => callback(obj), callbackState, TimeSpan.FromMilliseconds(-1), TimeSpan.FromMilliseconds(-1));
-        }
+        private Timer _timer;
+        private readonly Action<object> _callback;
+        private readonly object _callbackState;
+        private SemaphoreSlim _sem;
```
Comment (abhipsaMisra) [inline iothub/service/src/ModernDotNet/Common/IOThreadTimerSlim.cs:14]: very nitty - rename it to createTimeSemaphore?

**Review by abhipsaMisra — APPROVED** [review]

**Pull request merged**